### PR TITLE
[READY] Dashboard 2024 FCTT

### PR DESCRIPTION
**Pull request by mcgalliard**

### CODE REVIEW

File: common/faults/fault_config.json [hunk 1]
```diff
@@ -199,6 +199,15 @@
                     "time_to_latch": 10,
                     "time_to_unlatch": 1000,
                     "lcd_message": "APPS Brake Fault"
+                },
+                {
+                    "fault_name": "dual_pedal_act",
```
Comment (AdityaAsGithub) [inline common/faults/fault_config.json:204]: APPS_BRAKE was originally the fault meant to handle the dual pedal fault, so get rid of that fault if you are changing the name. You can also just change the message displayed on the screen for APPS_BRAKE

File: source/dashboard/main.c [hunk 2]
```diff
@@ -433,6 +432,94 @@ void heartBeatLED()
...
+    }
+    
+    // Check if EXTI10 triggered the interrupt
+    // PD9: Quad Encoder A
+    if (EXTI->PR & EXTI_PR_PR10) { 
+        // Clear the interrupt pending bit for EXTI10
+        EXTI->PR |= EXTI_PR_PR10; 
+
+        // Interrupt handling for quad here
+        encoder_ISR();
```
Comment (AdityaAsGithub) [inline source/dashboard/main.c:466]: Maybe call encoder_ISR before you clear the pending bit - if there is another interrupt while you are in this function you might get screwed up, so if the flag is still high so you won't get that issue.

**Review by AdityaAsGithub — CHANGES_REQUESTED** [review]
Looks good, just two tiny things

File: common/faults/fault_config.json (hunk 1 above)
Comment (mcgalliard) [reply common/faults/fault_config.json:204]: Will address

File: source/dashboard/main.c [hunk 3]
```diff
@@ -433,6 +432,114 @@ void heartBeatLED()
...
+    
+    // Check if EXTI10 triggered the interrupt
+    // PD9: Quad Encoder A
+    if (EXTI->PR & EXTI_PR_PR10) { 
+
+        // Interrupt handling for quad here
+        encoder_ISR();
+
+        // Clear the interrupt pending bit for EXTI10
+        EXTI->PR |= EXTI_PR_PR10;
```
Comment (LukeOxley) [inline source/dashboard/main.c:468]: You used = to clear interrupt on line 445 but |= here t

File: source/dashboard/main.c [hunk 4]
```diff
@@ -433,6 +432,114 @@ void heartBeatLED()
...
+        
+        EXTI->PR = EXTI_PR_PR9; // Clear the interrupt pending bit for EXTI9
+
+    }
+}
+
+void EXTI15_10_IRQHandler() {
+    // Check if the interrupt was caused by EXTI11
+    if (EXTI->PR & EXTI_PR_PR11) {
+        PHAL_toggleGPIO(PRCHG_LED_GPIO_Port, PRCHG_LED_Pin);
```
Comment (LukeOxley) [inline source/dashboard/main.c:453]: This line needs to go eventually

File: source/dashboard/main.c [hunk 5]
```diff
@@ -433,6 +432,114 @@ void heartBeatLED()
...
+        encoder_count--;
+    } else if ((previousA == 0 && previousB == 1) && (currentA == 1 && currentB == 1)) {
+        // Counterclockwise movement detected
+        encoder_count--;
+    } else if ((previousA == 0 && previousB == 0) && (currentA == 0 && currentB == 1)) {
+        // Counterclockwise movement detected
+        encoder_count--;
+    }
+
+    previousA = currentA;
```
Comment (LukeOxley) [inline source/dashboard/main.c:506]: Proposed method:

The switch only has 4 possible output states. You could convert from the gray code to the state [0,3].
Store the previous state as the static variable.
Then, on interrupt, decipher the gray code using a switch case to get the new position [0,3].

Update the tick count by the delta in position.

File: source/dashboard/main.c [hunk 6]
```diff
@@ -433,6 +432,114 @@ void heartBeatLED()
...
+    previousB = currentB;
+}
+
+void enableInterrupts()
```
Comment (LukeOxley) [inline source/dashboard/main.c:510]: Thx for moving this to its own function

**Comment by LukeOxley** [comment]
Found that the Start button press and page change buttons were adding items to can and uart queues from interrupt context. This has the potential to mess up the buffer if it preempts a buffer operation.

File: source/dashboard/main.c [hunk 7]
```diff
@@ -433,62 +373,156 @@ void heartBeatLED()
...
+    // EXTI14 triggered the interrupt (B1_FLT)
+    // This is the TOP button on the dashboard
+    if (EXTI->PR & EXTI_PR_PR14) {
+        if (sched.os_ticks - last_click_time < 300) {
+            last_click_time = sched.os_ticks;
+            EXTI->PR |= EXTI_PR_PR14;       // Clear the interrupt pending bit for EXTI14
+        }
+        else {
+            last_click_time = sched.os_ticks;
+            moveUp();
```
Comment (LukeOxley) [inline source/dashboard/main.c:405]: not interrupt safe

File: source/dashboard/main.c [hunk 8]
```diff
@@ -433,62 +373,156 @@ void heartBeatLED()
...
+    if (EXTI->PR & EXTI_PR_PR13)
+    {
+        if (sched.os_ticks - last_click_time < 300) {
+            last_click_time = sched.os_ticks;
+            EXTI->PR |= EXTI_PR_PR13;       // Clear the interrupt pending bit for EXTI13
+        }
+        else
+        {
+            last_click_time = sched.os_ticks;
+            moveDown();
```
Comment (LukeOxley) [inline source/dashboard/main.c:421]: not interrupt safe

File: source/dashboard/main.c [hunk 9]
```diff
@@ -433,62 +373,156 @@ void heartBeatLED()
...
+    if (EXTI->PR & EXTI_PR_PR12)
+    {
+        if (sched.os_ticks - last_click_time < 300) {
+            last_click_time = sched.os_ticks;
+            EXTI->PR |= EXTI_PR_PR12;       // Clear the interrupt pending bit for EXTI12
+        }
+        else
+        {
+            last_click_time = sched.os_ticks;
+            selectItem();
```
Comment (LukeOxley) [inline source/dashboard/main.c:437]: not interrupt safe

File: source/dashboard/main.c [hunk 10]
```diff
@@ -433,62 +373,156 @@ void heartBeatLED()
...
+            EXTI->PR |= EXTI_PR_PR12;       // Clear the interrupt pending bit for EXTI12
+        }
+    }
+
+    // EXTI11 triggered the interrupt (START_FLT)
+    if (EXTI->PR & EXTI_PR_PR11) {
+        PHAL_toggleGPIO(ERROR_LED_GPIO_Port, ERROR_LED_Pin); // Toggle LED for testing
+        SEND_START_BUTTON(q_tx_can, 1); // Report start button pressed
```
Comment (LukeOxley) [inline source/dashboard/main.c:445]: not interrupt safe

**Review by LukeOxley — CHANGES_REQUESTED** [review]

**Comment by LukeOxley** [comment]
Ensure the execution time of lcd stuff is not interfering with the 15ms critical system functions (pedals). It can't really exceed about 1ms or the watchdog will get mad.

**Review by AdityaAsGithub — APPROVED** [review]

**Pull request merged**